src/app/code/community/Shopgate/Framework/controllers/FrameworkController.php

```diff
@@ -83,17 +83,21 @@ protected function _run()
             $plugin  = Mage::getModel('shopgate/shopgate_plugin', $builder);
             $plugin->handleRequest(Mage::app()->getRequest()->getParams());
         } catch (ShopgateLibraryException $e) {
+            $traceId = Mage::app()->getRequest()->getParam('trace_id');
             $response = new ShopgatePluginApiResponseAppJson(
-                (isset($_REQUEST['trace_id']) ? $_REQUEST['trace_id'] : '')
+                (isset($traceId) ? $traceId : '')
             );
             $response->markError($e->getCode(), $e->getMessage());
             $response->setData(array());
             $response->send();
```

Comment:
without the `exit` here the response header is `application/text` instead of json. This happens because magento `Mage::app()->getResponse()` is being sent after this one. We encountered the same issue in Cloud plugin where I had to pick up all the headers from the API_Response class and rewrite the mage request class. Not sure how to handle this one as the formatting of the response happens in the `send()` function :/